### ISSUE TITLE
[Feature Request] Release a 24.04 version

### ISSUE THREAD

**Issue by gboutry** [canonical/self-signed-certificates-operator#270]
### Enhancement Proposal

This component is deployed in Sunbeam on an infrastructure VM that we'd like to move to 24.04 as default host (along with other components).

Would it be possible to also provide a release for 24.04?

**Comment by gruyaume** [comment]
Hello @gboutry,

Charmcraft, as of version `3` does not support having multiple bases. This is why, I believe, almost no charm supports both `22.04` and `24.04`. Here are the various options.

### Option 1: Replace `22.04` with `24.04` as the charm base

This is technically very simple, replace `22.04` with `24.04` in `charmcraft.yaml`. However, we want to keep on supporting 22.04 **and** we don't want to maintain a matrix of branches in every charm for:
- Arch
- Ubuntu Release
- Workload version

### Option 2: Support different charm bases in 1 repo

While `charmcraft.yaml` only supports 1 base, we could have different charmcraft<base>.yaml files but we'd need some hackery to build the right charm and the development experience would suffer a bit as charmcraft only looks for `charmcraft.yaml`. Because multi base support is coming (soon), I want to avoid building temporary hackery.

### Option 3: Wait for proper multi base support in charmcraft

As you can see in the references below, the *craft team is working on adding multi-base support to charmcraft. Here we can simply wait for this to be added to charmcraft and leverage it when it's there. As you can imagine, this is my preferred option.

## Reference
- https://juju.is/docs/sdk/charmcraft-yaml#heading--bases
- https://github.com/canonical/charmcraft/issues/2008
- https://docs.google.com/document/d/1QVHxZumruKVZ3yJ2C74qWhvs-ye5I9S6avMBDHs2YcQ/edit?tab=t.0

**Comment by gboutry** [comment]
At the moment, our goal is to release a candidate version before Christmas, where at that point no breaking change would land on edge or beta.

This charm being 22.04 forces one of the machines of the deployment to be in 22.04 while we're trying to move everything to 24.04. Alas, upgrading to 24.04 the machine on a running deployment is not a trivial undertaking.

**Comment by gruyaume** [comment]
@gboutry why not deploy `self-signed-certificates` on Kubernetes? As far as I know, the Vault Kubernetes charm is used in Sunbeam as well right?

I'm opened to other solutions by the way, if you can find a way to support multiple bases in a sustainable way.

**Comment by gboutry** [comment]
Because we deploy this charm before deploying K8S. It's used to secure communication from the sunbeam client to its state store.

Could it be possible to maintain a 24.04 on a separate branch for the time being and then move to single branch again when multiple base support is added?

**Comment by gruyaume** [comment]
We could, but then we're adding an overhead for the engineering team as we'll need to make changes on those 2 branches.

If we are to support theses multiple bases, I would favor a file-based approach and avoid branching altogether. 

@ghislainbourgeois can I get your opinion on this topic?

**Comment by ghislainbourgeois** [comment]
I think a file based approach would be easier to maintain going forward, it would mean that code fixes would not have to be back ported to another branch.

We should look into how other teams are handling that problem also, particularly around the CI to not reinvent the wheel.

**Comment by gboutry** [comment]
ubuntu-advantage charm is an example of a charm targeting 22.04 and 24.04

**Comment by gruyaume** [comment]
Yes, because they're not using charmcraft 3. For reference: https://git.launchpad.net/charm-ubuntu-advantage/tree/charmcraft.yaml

**Comment by ghislainbourgeois** [comment]
I thought only charmcraft 3 supports 24.04. looking at that charmcraft.yaml, it does not list 24.04, so maybe they use another file for charmcraft 3 for 24.04 support?

**Comment by gboutry** [comment]
They have another branch with a charmcraft.yaml 3 specific that publishes to the same track

**Comment by gruyaume** [comment]
As discussed with @gboutry , the Openstack team will fork the repo, do a sunbeam-ssc charm supporting 24.04 and cross-upgrade the charm back to self-signed-certificates when it supports 24.04. 

I'll keep this issue open until supporting multiple bases in the charm ecosystem becomes reasonable again and we can leverage the charmcraft feature.
